### PR TITLE
Implementing rainlang unit testing

### LINKED ISSUE #899: Unit testing

### PR DESCRIPTION
**Pull request by findolor**
Closes #899

## Motivation

See issue: #899

## Solution

- Added unit test config struct for mapping test configuration
- Added a new runner for the unit testing

Unit test rust tests can be run using:

```bash
nix develop -c cargo test unit_tests::tests::test_test_runner
```

## Checks

By submitting this for review, I'm confirming I've done the following:
- [x] made this PR as small as possible
- [x] unit-tested any new functionality
- [x] linked any relevant issues or PRs

### CODE REVIEW

File: crates/common/src/unit_tests/mod.rs [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+            settings: Settings {
+                main_config: settings,
+                test_config: test_settings,
+            },
+            rng: TestRng::from_seed(RngAlgorithm::ChaCha, &seed.unwrap_or([0; 32])),
+            test_setup: TestSetup {
+                block_number: 0,
+                deployer: Arc::new(Deployer {
+                    address: Address::default(),
+                    network: Arc::new(Network {
```
Comment (thedavidmeister) [inline crates/common/src/unit_tests/mod.rs:92]: if this is supposed to be some kind of dummy network, can it be an impl on `Network`? this logic looks like bad boilerplate here

File: crates/common/src/unit_tests/mod.rs [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+                decode_errors: true,
+            };
+            fork_clone
+                .fork_eval(args)
+                .map_err(TestRunnerError::ForkCallError)
+                .await
+        });
+
+        let result: RainEvalResults = vec![handle.await??.into()].into();
+        let flattened = result.into_flattened_table().unwrap();
```
Comment (thedavidmeister) [inline crates/common/src/unit_tests/mod.rs:182]: don't unwrap

File: crates/common/src/unit_tests/mod.rs [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+    }
+
+    async fn run_calculate_entrypoint(
+        &mut self,
+        pre_stack: Vec<U256>,
+    ) -> Result<RainEvalResults, TestRunnerError> {
+        let input_token = pre_stack[0];
+        let output_token = pre_stack[1];
+        let output_cap = pre_stack[2];
+        // let block_number = pre_stack[3];
```
Comment (thedavidmeister) [inline crates/common/src/unit_tests/mod.rs:193]: remove commented code

File: crates/common/src/unit_tests/mod.rs [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+        let dotrain = Arc::clone(&dotrain);
+
+        let handle = tokio::spawn(async move {
+            let rainlang_string =
+                RainDocument::compose_text(&dotrain, &["handle-io"], None, Some(final_bindings))?;
+
+            let mut context = vec![vec![U256::from(0); 5]; 5];
+
+            // output vault decrease
+            context[4][4] = max_output;
```
Comment (thedavidmeister) [inline crates/common/src/unit_tests/mod.rs:263]: doesn't it need to be the min of the calculated output and the vault balance set in pre?

File: crates/settings/src/unit_test.rs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        }
+
+        let scenario = Arc::new(Scenario {
+            name: self.scenario_name.clone(),
+            bindings: bindings.clone(),
+            runs: self.scenario.runs,
+            blocks: self.scenario.blocks.clone(),
+            deployer: Arc::new(Deployer {
+                address: Address::default(),
+                network: Arc::new(Network {
```
Comment (thedavidmeister) [inline crates/settings/src/unit_test.rs:53]: this should use the imple mentioned above

File: crates/common/src/unit_tests/mod.rs [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+    # calculate-entrypoint: some-custom-entrypoint
+    # handle-entrypoint: some-custom-entrypoint
+    scenario-name: some-key
+    scenario:
+        bindings:
+            orderbook-subparser: {orderbook_subparser}
+            second-binding: 999
+---
+#pre
+input-token: 0x0165878a594ca255338adfa4d48449f69242eb8f,
```
Comment (thedavidmeister) [inline crates/common/src/unit_tests/mod.rs:413]: do we have a test that this works, e.g. in post?

File: crates/common/src/unit_tests/mod.rs [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+output-token: 0xa513e6e4b8f2a923d98304ec87f64353c4d5c853,
```
Comment (thedavidmeister) [inline crates/common/src/unit_tests/mod.rs:414]: do we have a test that this works, e.g. in post?

File: crates/common/src/unit_tests/mod.rs [hunk 8]
```diff
@@ -0,0 +1,470 @@
...
+output-cap: 10,
```
Comment (thedavidmeister) [inline crates/common/src/unit_tests/mod.rs:415]: do we have a test that this works, e.g. in post?

File: crates/common/src/unit_tests/mod.rs [hunk 9]
```diff
@@ -0,0 +1,470 @@
...
+block-number: 100;
```
Comment (thedavidmeister) [inline crates/common/src/unit_tests/mod.rs:416]: do we have a test that this works, e.g. in post?

**Review by thedavidmeister — CHANGES_REQUESTED** [review]

File: crates/common/src/unit_tests/mod.rs (hunk 4 above)
Comment (thedavidmeister) [reply crates/common/src/unit_tests/mod.rs:250]: this is still relevant and a major bug, right?

**Review by thedavidmeister — CHANGES_REQUESTED** [review]

File: crates/common/src/unit_tests/mod.rs (hunk 4 above)
Comment (findolor) [reply crates/common/src/unit_tests/mod.rs:250]: @thedavidmeister you are right, fixing that now

**Review by thedavidmeister — CHANGES_REQUESTED** [review]

**Review by thedavidmeister — APPROVED** [review]

**Review by thedavidmeister — CHANGES_REQUESTED** [review]
failing ci

**Review by thedavidmeister — APPROVED** [review]

**Pull request merged**